pkg/services/cosmosdb/cosmos-utils.go

```diff
@@ -178,3 +182,308 @@ func deleteDatabase(
+		readLocations,
+		dt.DatabaseAccountName,
+	)
+	if pp.GetString("autoFailoverEnabled") == enabled {
+		p["enableAutomaticFailover"] = true
+	} else {
+		p["enableAutomaticFailover"] = false
+	}
+
+	filters := []string{}
+	ipFilters := pp.GetObject("ipFilters")
+	if ipFilters.GetString("allowAzure") == disabled &&
+		ipFilters.GetString("allowPortal") != disabled {
+		// Azure Portal IP Addresses per:
+		// https://aka.ms/Vwxndo
+		//|| Region            || IP address(es) ||
+		//||=====================================||
+		//|| China             || 139.217.8.252  ||
+		//||===================||================||
+		//|| Germany           || 51.4.229.218   ||
+		//||===================||================||
+		//|| US Gov            || 52.244.48.71   ||
+		//||===================||================||
+		//|| All other regions || 104.42.195.92  ||
+		//||                   || 40.76.54.131   ||
+		//||                   || 52.176.6.30    ||
+		//||                   || 52.169.50.45   ||
+		//||                   || 52.187.184.26  ||
+		//=======================================||
+		// Given that we don't really have context of the cloud
+		// we are provisioning with right now, use all of the above
+		// addresses.
+		filters = append(filters,
+			"104.42.195.92",
+			"40.76.54.131",
+			"52.176.6.30",
+			"52.169.50.45",
+			"52.187.184.26",
+			"51.4.229.218",
+			"139.217.8.252",
+			"52.244.48.71",
+		)
+	} else {
+		filters = append(filters, "0.0.0.0")
+	}
+	filters = append(filters, ipFilters.GetStringArray("allowedIPRanges")...)
+	if len(filters) > 0 {
+		p["ipFilters"] = strings.Join(filters, ",")
+	}
+	p["consistencyPolicy"] = pp.GetObject("consistencyPolicy").Data
+	return p, nil
+}
+
+type readLocationInformation struct {
+	ID       string
+	Location string
+	Priority int
+}
+
+func buildReadLocationInformation(
+	readLocations []string,
+	databaseAccountName string,
+) []readLocationInformation {
+	informations := []readLocationInformation{}
+	for i := range readLocations {
+		information := readLocationInformation{}
+		information.Location = readLocations[i]
+		information.Priority = i
+		information.ID = generateIDForReadLocation(
+			databaseAccountName,
+			readLocations[i],
+		)
+		informations = append(informations, information)
+	}
+	return informations
+}
+
+// Because the database account name is determined during provision step,
+// it is possible the database account name has length 36 (the length of UUID).
+// And in update step, a read location whose name is longer than 14 character
+// may be provided, in which case will cause bug in updating.
+// To avoid this case, we truncate the id of read locations to 50 characters.
+// It is tested that truncating the read region id won't affect user's usage.
+func generateIDForReadLocation(
+	databaseAccountName string,
+	location string,
+) string {
+	locationID := fmt.Sprintf("%s-%s", databaseAccountName, location)
+	if len(locationID) > 50 {
+		locationID = locationID[0:50]
```

Comment:
There is a simliar function `generateAccountName` in common-provision.go. Does it make more sense to follow a similar implementation, or just extend it and reuse?

Reply:
The reason is this function is invoked in both provisioning and updating, and if the region is not changed, we must guarantee the id is not changed. `generateAccountName ` will generate a new name every time it is invoked.